llvm/include/llvm/CASObjectFormats/SchemaBase.h

```diff
@@ -38,6 +38,11 @@ class SchemaBase {
   virtual void anchor();
 
 public:
+  /// Creates a specific \p SchemaBase implementation by inferring the schema
+  /// from a \p CASID of a root node of that schema.
+  static Expected<std::unique_ptr<SchemaBase>> create(cas::CASDB &CAS,
+                                                      cas::CASID ID);
+
```

Comment:
Two comments on this:
- It's encouraging an anti-pattern, since `SchemaBase` does work in its constructor and is immutable after creation. It should be reused.
- The naming / usage suggests that the `ID` is the schema.

Please refactor outside of `SchemaBase` into something like this:
```
class SchemaPool {
public:
  // Look up the schema for the node. Returns `*` if it's not in the schema.
  //
  // Thread-safe.
  SchemaBase *getSchemaFor(cas::NodeRef) const;

  // Not sure we need this wrapper, but allows a drop-in replacement for the usage you had before...
  Expected<SchemaBase *> getSchemaFor(cas::CASID ID) const {
    if (auto Node = CAS.getNode(ID))
      return getSchemaFor(*Node);
    else
      return Node.takeError();
  }

  // Creates all the schemas up front.
  SchemaPool(cas::CASDB &CAS);

private:
  SmallVector<std::unique_ptr<SchemaBase>> Schemas;
};
```
Later, we can:
- Add a `getSchema()` function that returns a schema by some sort of identifier for that schema.
- Make `SchemaPool` lazy (if we grow lots of schemas).

Also, it's fine if lld creates a separate SchemaPool for each object file (for now); it can be fixed later to share a single pool.

Reply:
Note that I think looking up a schema should have "optional" semantics in general, not "expected".
- In `lld`, you know from context that it should be part of a schema.
- In `llvm-cas-object-format`, you're doing this check for every node with no knowledge of context.

Reply:
Thank you for the suggestions! I'll look into using a pool pattern.

> Note that I think looking up a schema should have "optional" semantics in general, not "expected".

If we want to have
```
getSchemaFor(cas::CASID ID) const;
```
for simplicity of the caller (which I think is useful, `lld` shouldn't need to care what is the underlying CAS node type for the root CAS object), then return value needs to be `Expected`.

Reply:
I don't think `getSchemaFor(NodeRef)` should return `Expected`, since in general it's unexpected that a node has a schema, and the error is giving zero information that a `nullptr` doesn't already provide. (BTW, a better name is likely `getSchemaForPossibleRoot()`)

Your reply discusses `getSchemaFor(cas::CASID ID)`. I don't actually think anyone should use this function; let's just not define it. The reason is that any caller will need to immediately call `CASDB::getNode()` anyway, which is redundant with the call inside the function.

Your current code is:
```
  Expected<std::unique_ptr<SchemaBase>> Schema = SchemaBase::create(CAS, ID);
   if (auto E = Schema.takeError())
     return E;
   Expected<cas::NodeRef> Ref = CAS.getNode(ID);
   if (auto E = Ref.takeError())
     return E;
```

I suggest updating it to:
```
  // FIXME: Should pass in a SchemaPool instead of a CAS so the pool can be reused.
  SchemaPool Pool(CAS);

  Optional<cas::NodeRef> Ref;
  if (Error E = CAS.getNode(ID).moveInto(Ref))
   return E;
  SchemaBase *Schema = Pool.getSchemaForPossibleRoot(*Ref);
  if (!Schema)
    return createStringError("CAS object is not an object file");
```
I think is simpler since it separates concerns.

WDYT?

Reply:
>  The reason is that any caller will need to immediately call CASDB::getNode() anyway

It doesn't make much difference in practice but could you clarify why the callers need to do it?
AFAICT `lld` mainly needs a way to convert `CASID -> get [interface over symbolic info]` to do symbol queries, it doesn't *need* to traffic and/or use a `cas::NodeRef` directly.

E.g. like doing something like this in single function call:
```
Expected<std::unique_ptr<CASObjectReader>> getObjectReader(cas::CASDB &CAS, cas::CASID ID);
```

Reply:
`getObjectReader` SGTM as a follow-up!